---
title: "Functional Analysis - Normal lung versus lung with distal tumor"
author: "`r getOption('author')`"
date: "`r Sys.Date()`"
bibliography: bibliography.bib
params:
    bcbFile: "data/bcb_sub_lungnotumor_tumor.rda"
    resFile: "data/res_lungnotumor_tumor.rda"
    organism: "Mm"
    gspecies: "mmu"
    species: "mouse"
    lfc: 0.5849625
    outputDir: "."
---

```{r setup, cache=FALSE, message=FALSE}
# Organism params:
# Mouse (default)
#     organism: "Mm"
#     gspecies: "mmu"
#     species: "mouse"
# Human
#     organism: "Hs"
#     gspecies: "hsa"
#     species: "human"

# Organism database package
# help("enrichGO", "clusterProfiler")
orgDB <- paste("org", params$organism, "eg.db", sep = ".")

# Load the required libraries
packages <- c(
    orgDB,
    "lpantano/DEGreport",
    "hbc/bcbioRNASeq",
    "tidyverse",
    "DT",
    "clusterProfiler",
    "DOSE",
    "pathview"
)
if (!all(basename(packages) %in% rownames(installed.packages()))) {
    source("https://bioconductor.org/biocLite.R")
    notInstalled <- setdiff(basename(packages), rownames(installed.packages()))
    biocLite(pkgs = notInstalled)
}
invisible(lapply(basename(packages), library, character.only = TRUE))
    
# Shared R Markdown settings
prepareRNASeqTemplate()
if (file.exists("setup.R")) {
    source("setup.R")
}

# Load bcbioRNASeq object
bcbName <- load(params$bcbFile)
bcb <- get(bcbName, inherits = FALSE)

resName <- load(params$resFile)
res <- get(resName, inherits = FALSE) %>%
    .[!is.na(.$padj), , drop = FALSE]
alpha <- metadata(res)$alpha

# Directory paths
faDir <- file.path(params$outputDir, "results", "functional_analysis")
dataDir <- file.path(params$outputDir, "data")
lapply(list(faDir, dataDir), dir.create, recursive = TRUE, showWarnings = FALSE) %>% invisible()
```

```{r header, child="_header.Rmd", eval=file.exists("_header.Rmd")}
```


```{r identifiers, message=FALSE, warning=FALSE}
allGenes <- rownames(res) %>%
    as.character()
sigGenes <- significants(
    res,
    padj = alpha, 
    fc <- params$lfc)
# Convert Ensembl gene identifiers to Entrez
ensembl2entrez <- bitr(
    allGenes,
    fromType = "ENSEMBL",
    toType = "ENTREZID",
    OrgDb = orgDB) %>%
    camel()
allEntrez <- unique(ensembl2entrez$entrezid)
sigEntrez <- bitr(
    sigGenes,
    fromType = "ENSEMBL",
    toType = "ENTREZID",
    OrgDb = orgDB) %>%
    camel() %>%
    .$entrezid %>%
    unique()
```

```{r sig_tables}
sigResults <- as.data.frame(res)[sigGenes, ]

foldChanges <- sigResults$log2FoldChange
names(foldChanges) <- rownames(sigResults)

signedpvalue_ordered <- res %>% 
  as.data.frame %>% 
  tibble::rownames_to_column(., "rownames") %>% 
  mutate(signedpvalue=-log(pvalue)*sign(log2FoldChange)) %>% 
  column_to_rownames(., "rownames") %>% 
  .[order(.$signedpvalue),]

pvalue_ordered <- res %>% 
  as.data.frame %>% 
  tibble::rownames_to_column(., "rownames") %>% 
  mutate(logpvalue=-log(pvalue)) %>% 
  column_to_rownames(., "rownames") %>% 
  .[order(.$logpvalue),]

```

# GO enrichment analysis {.tabset}

Gene Ontology (GO) term enrichment is a technique for interpreting sets of genes making use of the Gene Ontology system of classification, in which genes are assigned to a set of predefined bins depending on their functional characteristics.

In an enrichment analysis, we take a set of differentially expressed (DE) genes and check to see if any of the gene sets (here from the GO sets) are found in our set of DE genes than you would expect at random. *If you want to get into the math, the main statistic used for these types of analyses is a [hypergeometic test](https://en.wikipedia.org/wiki/Hypergeometric_distribution#Hypergeometric_test)

There are 3 different domains in the GO tree: Biological processes (often the most relevant), cellular component (i.e. where the genes' proteins are located) and molecular function (i.e. enzymatic function). We look at them separately here using only the genes with an adjusted pvalue of less than `r alpha` and at least an absolute log2FC of `r params$lfc` as our set of differentially expressed genes (these cutoffs are fairly abritrary, but were chosen to give us a set of genes taht were significant and had at least some change in expression). A GO category is considered enriched if the statistical test returns an adjusted pvalue (for multiple testing) of less than `r alpha`.

## Biological Processes

```{r enrich_goBP}
# Run GO enrichment analysis
egoBP <- enrichGO(
    sigGenes,
    "ENSEMBL",
    universe = allGenes,
    OrgDb = orgDB,
    ont = "BP",
    pAdjustMethod = "BH",
    qvalueCutoff = alpha,
    readable = TRUE)
egoBP_lungnotumor_tumor <- egoBP
saveData(egoBP_lungnotumor_tumor, dir = dataDir)

# Show summary data.frame
egoBPSummary <- slot(egoBP, "result") %>%
    as_tibble() %>%
    camel()
write_csv(
    egoBPSummary,
    path = file.path(
        faDir,
        paste0(
            paste("go", 
                  "BP",
                  "clusterprofiler",
                  "lungnotumor_tumor",
                  "padj",
                  alpha,
                  "lfc",
                  params$lfc,
                  sep = "_"),
            ".csv.gz")))


egoBPSummary %>% datatable()
```

## Cellular Component

```{r enrich_goCC}
# Run GO enrichment analysis
egoCC <- enrichGO(
    sigGenes,
    "ENSEMBL",
    universe = allGenes,
    OrgDb = orgDB,
    ont = "CC",
    pAdjustMethod = "BH",
    qvalueCutoff = alpha,
    readable = TRUE)
saveData(egoCC_lungnotumor_tumor, dir = dataDir)

# Show summary data.frame
egoCCSummary <- slot(egoCC, "result") %>%
    as_tibble() %>%
    camel()
write_csv(
    egoCCSummary,
    path = file.path(
        faDir,
        paste0(
            paste("go", 
                  "CC",
                  "clusterprofiler",
                  "lungnotumor_tumor",
                  "padj",
                  alpha,
                  "lfc",
                  params$lfc,
                  sep = "_"),
            ".csv.gz")))


egoCCSummary %>% datatable()
```

## Molecular Function

```{r enrich_goMF}
# Run GO enrichment analysis
egoMF <- enrichGO(
    sigGenes,
    "ENSEMBL",
    universe = allGenes,
    OrgDb = orgDB,
    ont = "MF",
    pAdjustMethod = "BH",
    qvalueCutoff = alpha,
    readable = TRUE)
saveData(egoMF_lungnotumor_tumor, dir = dataDir)

# Show summary data.frame
egoMFSummary <- slot(egoMF, "result") %>%
    as_tibble() %>%
    camel()
write_csv(
    egoMFSummary,
    path = file.path(
        faDir,
        paste0(
            paste("go", 
                  "MF",
                  "clusterprofiler",
                  "lungnotumor_tumor",
                  "padj",
                  alpha,
                  "lfc",
                  params$lfc,
                  sep = "_"),
            ".csv.gz")))


egoMFSummary %>% datatable()
```


# DOTplot {.tabset}

## Biological Process

```{r dotplotBP, fig.width=8}
# Dotplot of top 25
dotplot(egoBP, showCategory = 25)
```

## Cellular Componenet

```{r dotplotcc, fig.width=8}
# Dotplot of top 25
dotplot(egoCC, showCategory = 25)
```

## Molecular Function

```{r dotplotmf, fig.width=8}
# Dotplot of top 25
dotplot(egoMF, showCategory = 25)
```

# GO terms map {.tabset}

## Biological Process

```{r enrich_mapbp, fig.width=8, fig.height=8}
# Enrichment plot of top 25
enrichMap(egoBP, n = 50, vertex.label.cex = 0.5)
```

## Cellular Componenet

```{r enrich_mapcc, fig.width=8, fig.height=8}
# Enrichment plot of top 25
enrichMap(egoCC, n = 50, vertex.label.cex = 0.5)
```

## Molecular Function

```{r enrich_mapmf, fig.width=8, fig.height=8}
# Enrichment plot of top 25
enrichMap(egoMF, n = 50, vertex.label.cex = 0.5)
```

# Gene map {.tabset}

In order to consider the potentially biological complexities in which a gene may belong to multiple annotation categories and provide information of numeric changes if available

## Biological Process

```{r cnetplotbp, fig.width=8, fig.height=8}
# Cnet plot with genes colored by fold changes for top 5 most significant GO
# processes
cnetplot(
    egoBP,
    categorySize = "pvalue",
    showCategory = 10,
    foldChange = foldChanges,
    vertex.label.cex = 0.5)
```

## Cellular Componenet

```{r cnetplotcc, fig.width=8, fig.height=8}
# Cnet plot with genes colored by fold changes for top 5 most significant GO
# processes
cnetplot(
    egoCC,
    categorySize = "pvalue",
    showCategory = 10,
    foldChange = foldChanges,
    vertex.label.cex = 0.5)
```

## Molecular Function

```{r cnetplotmf, fig.width=8, fig.height=8}
# Cnet plot with genes colored by fold changes for top 5 most significant GO
# processes
cnetplot(
    egoMF,
    categorySize = "pvalue",
    showCategory = 10,
    foldChange = foldChanges,
    vertex.label.cex = 0.5)
```

# KEGG analysis

Here we do the same type of enrichment analysis, but using the genes in KEGG defined pathways as gene sets instead. We use the same definition of DE genes.

```{r enrich_kegg}
kegg <- enrichKEGG(
    gene = sigEntrez,
    universe = allEntrez,
    organism = params$species, 
    pvalueCutoff = alpha)
saveData(kegg_lungnotumor_tumor, dir = dataDir)

# Show KEGG summary data.frame
keggSummary <- slot(kegg, "result") %>%
    as_tibble() %>%
    camel()
write_csv(
    keggSummary,
    path = file.path(
        faDir,
        paste0(
            paste(
                "kegg",
                "clusterprofiler",
                "lungnotumor_tumor",
                "padj",
                alpha,
                "lfc",
                params$lfc,
                sep = "_"),
            ".csv.gz")))
keggSummary %>% datatable()
```


# GO GSEA analysis

A common approach in analyzing gene expression profiles was identifying differential expressed genes that are deemed interesting. The enrichment analysis we demonstrated previously were based on these differentially expressed genes. This approach will find genes where the difference is large, but it will not detect a situation where the difference is small, but evidenced in coordinated way in a set of related genes. Gene Set Enrichment Analysis (GSEA) directly addresses this limitation. All genes can be used in GSEA; GSEA aggregates the per gene statistics across genes within a gene set, therefore making it possible to detect situations where all genes in a predefined set change in a small but coordinated way. 

GSEA relies on ranking genes. Choosing a metric to rank by is an important decision which affects the interpretation of the results. Here we rank them by a metric based on their -log(pvalues), but modified by the sign of the log2FC (i.e. upregulated and downregulated genes will be put at opposite ends of the ranking). This will result in a gene ranking where the significance of the gene expression change determines ranking and negatively and postively regulated genes are treated as being different. 

```{r gse_gobp_signedpvalue}
# Prepare the gene list
geneList <- signedpvalue_ordered$signedpvalue
names(geneList) <- row.names(signedpvalue_ordered)
geneList <- geneList[!is.na(geneList)]

# Now run GSEA
gsesignedpvalue <- gseGO(
    geneList = rev(geneList),
    OrgDb = orgDB,
    ont = "ALL",
    keyType = "ENSEMBL",
    nPerm = 1000,
    minGSSize = 10,
    maxGSSize = 500,
    pvalueCutoff = alpha,
    verbose = FALSE,
    by="fgsea")
saveData(gsesignedpvalue_lungnotumor_tumor, dir = dataDir)

# Write out summary data.frame
gsesignedpvalueSummary <- slot(gsesignedpvalue, "result") %>%
    as_tibble() %>%
    camel()
write_csv(
    gsesignedpvalueSummary,
    path = file.path(
        faDir,
        paste0(
            paste(
                "gsea",
                "clusterprofiler",
                "lungnotumor_tumor",
                "padj",
                alpha,
                "lfc",
                params$lfc,
                sep = "_"),
            ".csv.gz")))
gsesignedpvalueSummary %>% datatable()
```

There is nothing enriched when we treat the direction of expression important. This may be because GO categories are not necessarily sets of coregulated genes, and the method we used will find sets of genes that are upregulated together OR downregulated together, rather than both.

```{r gse_gobp_pvalue}
# Prepare the gene list
geneList <- pvalue_ordered$logpvalue
names(geneList) <- row.names(pvalue_ordered)
geneList <- geneList[!is.na(geneList)]

# Now run GSEA
gsepvalue <- gseGO(
    geneList = rev(geneList),
    OrgDb = orgDB,
    ont = "ALL",
    keyType = "ENSEMBL",
    nPerm = 1000,
    minGSSize = 10,
    maxGSSize = 500,
    pvalueCutoff = alpha,
    verbose = FALSE,
    by="fgsea")
saveData(gsepvalue_lungnotumor_tumor, dir = dataDir)

# Write out summary data.frame
gsepvalueSummary <- slot(gsepvalue, "result") %>%
    as_tibble() %>%
    camel()
write_csv(
    gsespvalueSummary,
    path = file.path(
        faDir,
        paste0(
            paste(
                "gsea",
                "clusterprofiler",
                "lungnotumor_tumor",
                "padj",
                alpha,
                "lfc",
                params$lfc,
                sep = "_"),
            ".csv.gz")))
gsepvalueSummary %>% datatable()
```

# KEGG GSEA analysis

We can also perform GSEA analysis with clusterProfiler using KEGG gene sets and using the log2 fold changes as input. By using the log2 fold changes as the input, we are identifying pathways with genes that exhibit coordinated fold changes that are larger than might be expected by chance. The significant pathways can be visualized using the log2 fold changes with the Pathview tool.

```{r kegg_gsea}
resTblEntrez <- res %>%
    as.data.frame() %>%
    rownames_to_column("ensembl") %>%
    as_tibble() %>%
    # Use the full join option here to include all Entrez identifiers
    full_join(ensembl2entrez, by = "ensembl") %>%
    dplyr::select(entrezid, ensembl, everything()) %>%
    # Remove rows with duplicate Entrez identifiers
    filter(!duplicated(entrezid))

# Extract the fold changes
foldChanges <- resTblEntrez$log2FoldChange
names(foldChanges) <- resTblEntrez$entrezid
foldChanges <- sort(foldChanges, decreasing = TRUE)

# GSEA using gene sets from KEGG pathways
gseaKEGG <- gseKEGG(
    geneList = foldChanges,
    organism = tolower(params$gspecies),
    nPerm = 1000,
    minGSSize = 20,
    pvalueCutoff = 0.05,
    verbose = FALSE)
saveData(gseKEGG_lungnotumor_tumor, dir = dataDir)

# Extract the GSEA results
gseaKEGGSummary <- slot(gseaKEGG, "result") %>% as_tibble()
write_csv(
    gseaKEGGSummary,
    path = file.path(faDir, "gsea_kegg_clusterprofiler_lungnotumor_tumor.csv.gz"))
gseaKEGGSummary %>% datatable()
```

```{r kegg_plots, message=FALSE, results='asis'}
detach("package:bcbioRNASeq", unload=TRUE)
# dplyr must be unloaded at this step for pathview to work
suppressWarnings(detach("package:dplyr", unload = TRUE))

# If there is an error at this step, there may be a pathway that is not found by
# pathview package. In this case, you may need to run the pathview command above
# by specifying the index of the pathways you would like to print out in place
# of `x`.
pathways <- gseaKEGGSummary$ID

keggPlotsDir <- file.path(faDir, "kegg_plots_lungnotumor_tumor")
dir.create(keggPlotsDir, recursive = TRUE, showWarnings = FALSE)

# There is currently no way to set the output path of the pathview PNG files.
# We're using tryCatch here to return to the user pathways that didn't output
# graphics correctly.
current <- getwd() %>% normalizePath()
setwd(keggPlotsDir)
lapply(seq_along(pathways), function(a) {
    pathway <- pathways[[a]]
    tryCatch(
        pathview(
            gene.data = foldChanges,
            pathway.id = pathway,
            species = tolower(params$gspecies), 
            limit = list(gene = 2, cpd = 1)),
        error = function(e) {
            warning(paste(pathway, "failed to plot"), call. = FALSE)
        }
    )
}) %>%
    invisible()
setwd(current)

figures <- list.files(keggPlotsDir, pattern = "pathview", full.names = TRUE)
invisible(lapply(seq_along(figures), function(a) {
    figure <- figures[[a]]
    cat(paste0("<img src=\"", figure, "\">\n"))
}))
```



```{r footer, child="_footer.Rmd", eval=file.exists("_footer.Rmd")}
```
