---
title: "Quality Control"
author: "`r getOption('author')`"
date: "`r Sys.Date()`"
bibliography: bibliography.bib
params:
    bcbFile: "data/bcb.rda"
    outputDir: "."
---

```{r setup, message=FALSE}
library(bcbioRNASeq)

# Shared R Markdown settings
prepareRNASeqTemplate()
if (file.exists("setup.R")) {
    source("setup.R")
}

# Load bcbioRNASeq object
bcbName <- load(params$bcbFile)
bcb <- get(bcbName, inherits = FALSE)
interestingGroups(bcb) <- "sampleclass"
# Directory paths
outputDir <- params$outputDir
dataDir <- file.path(outputDir, "data")
countsDir <- file.path(outputDir, "results", "counts")
```

```{r header, child="_header.Rmd", eval=file.exists("_header.Rmd")}
```



```{r sample_metadata, results='asis'}
sampleMetadata(bcb)
```

[bcbio][] run data was imported from **`r metadata(bcb)[["uploadDir"]]`**.

```{r counts}
rawCounts <- counts(bcb, normalized = FALSE)
normalizedCounts <- counts(bcb, normalized = TRUE)
tpm <- tpm(bcb)
saveData(rawCounts, normalizedCounts, tpm, dir = dataDir)
writeCounts(rawCounts, normalizedCounts, tpm, dir = countsDir)


counts <- counts(bcb, normalized = FALSE)
rowData <- rowData(bcb)
annotated <- cbind(as.data.frame(counts), as.data.frame(rowData))

```



# Read metrics {.tabset}

## Total reads

```{r plot_total_reads}
plotTotalReads(bcb)
```

The number of reads per sample look acceptable (i.e. above 10 million reads).

## Mapped reads

The number of mapped reads should correspond to the number of total reads.

```{r plot_mapped_reads}
plotMappedReads(bcb)
```

Still looking good.

## Mapping rate

The genomic mapping rate represents the percentage of reads mapping to the reference genome. Low mapping rates are indicative of sample contamination, poor sequencing quality or other artifacts.

```{r plot_mapping_rate}
plotMappingRate(bcb)
bcb <- bcb[,which(!(colData(bcb)$sampleID %in% c("lung_tumorbearing_6884_2", "lung_tumorbearing_6881_2", "lung_tumorbearing_6878_2")))]
```

The mapping rate is too low on 3 of the lungs with tumor present. I removed these samples now.

## Number of genes detected

```{r plot_genes_detected}
plotGenesDetected(bcb)
```

The gene detection rate is acceptable for all samples, but a bit lower for 1  samples (two lung samples with tumor present and  a primary tumor)

## Gene detection saturation

We should observe a linear trend in the number of genes detected with the number of mapped reads, which indicates that the sample input was not overloaded.

```{r plot_gene_saturation}
plotGeneSaturation(bcb)
```

Generally, gene detection looks saturated, with the exception of 3 samples.

## Exonic mapping rate

Ideally, at least 60% of total reads should map to exons.

```{r plot_exonic_mapping_rate}
plotExonicMappingRate(bcb)
```

Everything looks fine here, the majority of reads map to exons.

## Intronic mapping rate

The majority of reads should map to exons and not introns.

```{r plot_intronic_mapping_rate}
plotIntronicMappingRate(bcb)
```

Everything looks fine here, with a lower than critical level of intronic mapping

## rRNA mapping rate

Samples should have a ribosomal RNA (rRNA) contamination rate below 10%.

```{r plot_rrna_mapping_rate}
plotRRNAMappingRate(bcb)
```

Ribosomal rates look  good.

## 5'->3' bias

```{r plot_53_bias}
plot53Bias(bcb)
# add 53bias to colData for later plotting
mycoldata <- as.data.frame(colData(bcb)) %>% tbl_df()
mymetrics <- as.data.frame(metadata(bcb)$metrics) %>% tbl_df()
newcoldata <- inner_join(mycoldata, mymetrics) %>% as.data.frame()
row.names(newcoldata) <- newcoldata$sampleID
identical(row.names(newcoldata), row.names(as.data.frame(colData(bcb))))
colData(bcb)$x53Bias <- newcoldata$x53Bias
# put it as a factor so that you can actually plot it, but save it as numeric so you can use for subsetting
colData(bcb)$x53Bias_factor <- as.factor(colData(bcb)$x53Bias)

```

We're seeing quite variable 5'-3' biases, with all except one control sample (6890_2) showing little bias. About two tumors (6887_2 and 6881_2) and 3 lungs (6884_2, 6887_1, 6887_2) in the presence of tumors show little evidence of bias. Unfortunately, 2 of the lungs in the presence of primary tumors that show little bias are from the same original tissue sample. This should still give us enough to work with however.

## Counts per gene

Generally, we expect similar count spreads for all genes between samples unless the library sizes or total RNA expression are different. The log10 TMM-normalized counts per gene normalization method [@Robinson:2010dd] equates the overall expression levels of genes between samples under the assumption that the majority of them are not differentially expressed. Therefore, by normalizing for total RNA expression by sample, we expect the spread of the log10 TMM-normalized counts per gene to be similar for every sample.

```{r plot_counts_per_gene}
plotCountsPerGene(bcb)
```

There are definitely some more variable samples here than we'd like to see (lung_tumor_6878_2 is higher than others of it's class for example, likely because it's not degraded)


## Count density

Generally, we expect similar count spreads for all genes between samples unless the total expressed RNA per sample is different.

```{r plot_count_density}
plotCountDensity(bcb)
```



# Fit modeling

Several quality metrics are first assessed to explore the fit of the model, before differential expression analysis is performed.


## Variance stabilization

The plots below show the standard deviation of normalized counts (`normalized_counts`) using `log2()`, `rlog()`, and variance stabilizing (`vst()`) transformations by `rank(mean)`. The transformations greatly reduce the standard deviation, with `rlog()` stabilizing the variance best across the mean.

```{r plot_mean_sd, fig.height=18, fig.width=6}
plotMeanSD(bcb)
```


## Dispersion

The following plot shows the dispersion by mean of normalized counts. We expect the dispersion to decrease as the mean of normalized counts increases.

```{r plot_disp_ests, fig.height=8, fig.width=8}
plotDispEsts(bcb)
```



# Sample similarity analysis

Before performing similarity analysis, we transform counts to log2, which acts to minimize large differences in sequencing depth and helps normalize all samples to a similar dynamic range. For RNA-seq count data, variance increases with the mean. Logarithmic transformation of normalized count values with a small pseudocount will account for large variations seen between the highest expressing genes so that these genes won't dominate the PCA plots. However, due to the strong noise among low count values due to Poisson, the general log2 transformation will amplify this noise, and instead, low count genes will now dominate the PCA plots. So instead, we use a transformation (`rlog()`) that gives similar results for high counts as a log2 transformation but also shrinks the values of low counts towards the genesâ€™ average across samples. We do this with the `rlog()` function in the [DESeq2][] package [@DESeq2], which we will later use for differential gene expression analysis.


## Principal component analysis (PCA)

PCA [@Jolliffe:2002wx] is a multivariate technique that allows us to summarize the systematic patterns of variations in the data. PCA takes the expression levels for genes and transforms it in principal component space, reducing each sample into one point. Thereby, we can separate samples by expression variation, and identify potential sample outliers. The PCA plot is a way to look at how samples are clustering.

```{r plot_pca, fig.height=12, fig.width=10}
plotPCA(bcb, interestingGroups=c("sampleclass"))
plotPCA(bcb, interestingGroups=c("rin"))
plotPCA(bcb, interestingGroups=c("x53Bias_factor"))
```
 
The 1st principal component is capturing the lung versus tumor variation. The second principal component appears to be capturing RNA quality (as evidenced by RIN and 5->3' bias measures). The relationship between RIN and 5'->3' bias is not straightforward however.

Looking at the gene coverage plots from the [multiqc report](multiqc_report.html), we can see that some samples show a clear "hump" in the middle of the transcripts. There is also a strange effect where there is more gene counts on chr4 and chr12 than you'd expect, with the coverage on chr4 roughly corresponding to the RIN of(and presencd of a hump) the bad samples. 

Sometimes these issues/metrics are the result of effects on one or two highly expressed genes. I looked at the highest expressed genes in the worst samples and noticed a lot of pseudogenes, ribosomal genes and miRNAs. So I decided to see the effects of dropping any genes that were not either a) protein coding or b)lincRNAs.

```{r plot_pca_sub, fig.height=12, fig.width=10}
bcb_sub <- bcb[which(annotated[,"biotype"]=="protein_coding"|annotated[,"biotype"]=="lincRNA"),]
plotPCA(bcb_sub, interestingGroups=c("sampleclass"))
plotPCA(bcb_sub, interestingGroups=c("rin"))
plotPCA(bcb_sub, interestingGroups=c("x53Bias_factor"))

bcb_sub <- bcb_sub[,which(colData(bcb_sub)$sampleID!="lung_normal_NA_6890_2")]
```

The samples no longer cluster as strongly by RIN or 5'->3' bias. The single outlier is a control lung sample that had a low RIN and 5'->3' bias, but did not possess the gene coverage hump or high chr4 counts. 

We have lots of control samples, so removing it seems fine.

```{r plot_pca_sub2, fig.height=12, fig.width=10}
plotPCA(bcb_sub, interestingGroups=c("sampleclass"))
plotPCA(bcb_sub, interestingGroups=c("rin"))
plotPCA(bcb_sub, interestingGroups=c("x53Bias_factor"))
```

The vast majority of the variation we are visualizing is from the differences between lungs and tumors. We can further subset the data to only lung samples to see how the data clusters by lung type.

```{r plot_pca_sublung, fig.height=12, fig.width=10}
bcb_sub_lung <- bcb_sub[,which(colData(bcb_sub)$tissue=="lung")]
plotPCA(bcb_sub_lung, interestingGroups=c("sampleclass"), label=TRUE)
plotPCA(bcb_sub_lung, interestingGroups=c("rin"))

```
 
While the clustering is unclear, this is actually looking more promising. Now we can actually start to see a batch effect, with PC1 generally capturing the preparation date. This is slightly correlated wtih RIN value, but not completely.
 
 **My plan at this point is to drop the really low RIN value samples (i.e. the 2.5s), then collapse the replicate values to adjust for batch to see how the data looks. 

## Covariates correlation with PCs

When multiple factors may influence the results of a given experiment, it is useful to assess which of them is responsible for the most variance as determined by PCA. We adapted the method described by Daily et al. where they integrated a method to correlate covariates with principal components values to determine the importance of each factor.

Here we are showing the correlational analysis of the rlog transformed count data's principal components with the metadata covariates of interest. Significant correlations (FDR < 0.1) are shaded from blue (anti-correlated) to orange (correlated), with non-significant correlations shaded in gray.

```{r plot_pca_covariates}
plotPCACovariates(bcb_sub)
```

We can do the same thing with the subsetted  to lung dataset

```{r plot_pca_covariates_sub}
plotPCACovariates(bcb_sub_lung)
```


## Hierarchical clustering

Inter-correlation analysis (ICA) is another way to look at how well samples cluster by plotting the correlation between the expression profiles of the samples.

```{r plot_correlation_heatmap, fig.height=8, fig.width=10}
plotCorrelationHeatmap(bcb_sub)
```

We can repeat this with the subsetted to lung dataset as well.

```{r plot_correlation_heatmap_sub, fig.height=8, fig.width=10}
plotCorrelationHeatmap(bcb_sub_lung)
```


## Collapsing batch

There is a pretty clear separation by batch at this point. While we can use easily regress out the batch effect when doing the DE in this case, I'd like to also use RUVseq to remove the batch ahead of time and then visualize the data to make sure the RIN value isn't confounded with sample class.


```{r footer, child="_footer.Rmd", eval=file.exists("_footer.Rmd")}
```
